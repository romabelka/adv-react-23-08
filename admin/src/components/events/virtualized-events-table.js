import React, { Component } from 'react'
import { connect } from 'react-redux'
import { InfiniteLoader, Table, Column } from 'react-virtualized'
import {
  fetchLazy,
  eventListSelector,
  loadedSelector,
  loadingSelector,
  toggleSelect as handleSelect
} from '../../ducks/events'
import Loader from '../common/loader'
import 'react-virtualized/styles.css'

export class EventsTable extends Component {
  static propTypes = {}

  constructor(props) {
    super(props)
    this.remoteRowCount = 298 //TODO hardcode
  }

  componentDidMount() {
    this.props.fetchLazy()
  }

  render() {
    if (this.props.loading && !this.props.loaded) return <Loader />
    const { events } = this.props
    console.log('events', events)

    return (
      <InfiniteLoader
        isRowLoaded={this.isRowLoaded}
        loadMoreRows={this.loadMoreRows}
        rowCount={this.remoteRowCount}
      >
        {({ onRowsRendered, registerChild }) => (
          <Table
            ref={registerChild}
            rowCount={events.length}
            rowGetter={this.rowGetter}
            rowHeight={40}
            headerHeight={50}
            overscanRowCount={5}
            width={700}
            height={300}
            onRowClick={this.handleRowClick}
            onRowsRendered={onRowsRendered}
            rowRenderer={this.getRowRenderer}
          >
            <Column dataKey="title" width={200} label="Title" />
            <Column dataKey="when" width={100} label="Date" />
            <Column dataKey="where" width={200} label="Place" />
          </Table>
        )}
      </InfiniteLoader>
    )
  }

  isRowLoaded = ({ index }) => index < this.props.events.length

  loadMoreRows = () => {
    this.props.fetchLazy()
  }

  handleRowClick = ({ rowData }) => this.props.handleSelect(rowData.id)

  rowGetter = ({ index }) => this.props.events[index]
}

export default connect(
  (state) => ({
    events: eventListSelector(state),
    loading: loadingSelector(state),
    loaded: loadedSelector(state)
  }),
  { fetchLazy, handleSelect }
)(EventsTable)
