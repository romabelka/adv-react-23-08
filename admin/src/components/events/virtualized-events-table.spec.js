import React from 'react'
import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { EventsTable } from './virtualized-events-table'
import Loader from '../common/loader'
import rawEvents from '../../mocks/conferences'

Enzyme.configure({ adapter: new Adapter() })

const events = rawEvents.slice(0, 10).map((event, id) => ({ id, ...event }))

describe('Virtualized EventsTable', () => {
  it('should display loader', () => {
    const container = shallow(<EventsTable loading />, {
      disableLifecycleMethods: true
    })

    expect(container.contains(<Loader />))
  })

  // because we can't actually simulate scroll and test that table renders all events from first to last, I've decided to do
  // a snapshot test; I've chosen the `render` renderer because its output reflects the exact data contained in the events list
  it('should match the snapshot on render', () => {
    const container = render(<EventsTable events={events} />, {
      disableLifecycleMethods: true
    })

    expect(container).toMatchSnapshot()
  })

  it('should fetch all events on mount', (done) => {
    shallow(<EventsTable events={[]} fetchAllEvents={done} />)
  })

  it('should select an event', () => {
    let selectedId = null
    const container = mount(
      <EventsTable
        events={events}
        handleSelect={(id) => (selectedId = id)}
        fetchAllEvents={() => {}}
      />
    )

    container
      .find('.ReactVirtualized__Table__row')
      .first(0)
      .simulate('click')

    expect(selectedId).toEqual(events[0].id)
  })
})
