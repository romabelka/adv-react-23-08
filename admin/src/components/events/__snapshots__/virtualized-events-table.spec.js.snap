// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Virtualized EventsTable should match the snapshot on render 1`] = `
<div
  class="ReactVirtualized__Table"
  role="grid"
>
  <div
    class="ReactVirtualized__Table__headerRow"
    role="row"
    style="height:80px;overflow:hidden;padding-right:0;width:500px"
  >
    <div
      class="ReactVirtualized__Table__headerColumn"
      role="columnheader"
      style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px"
    >
      <span
        class="ReactVirtualized__Table__headerTruncatedText"
        title="Title"
      >
        Title
      </span>
    </div>
    <div
      class="ReactVirtualized__Table__headerColumn"
      role="columnheader"
      style="flex:0 1 100px;-ms-flex:0 1 100px;-webkit-flex:0 1 100px"
    >
      <span
        class="ReactVirtualized__Table__headerTruncatedText"
        title="Date"
      >
        Date
      </span>
    </div>
    <div
      class="ReactVirtualized__Table__headerColumn"
      role="columnheader"
      style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px"
    >
      <span
        class="ReactVirtualized__Table__headerTruncatedText"
        title="Place"
      >
        Place
      </span>
    </div>
  </div>
  <div
    aria-label="grid"
    aria-readonly="true"
    class="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
    role="rowgroup"
    style="box-sizing:border-box;direction:ltr;height:320px;position:relative;width:500px;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:auto"
    tabindex="0"
  >
    <div
      class="ReactVirtualized__Grid__innerScrollContainer"
      role="rowgroup"
      style="width:auto;height:500px;max-width:500px;max-height:500px;overflow:hidden;pointer-events:;position:relative"
    >
      <div
        aria-label="row"
        class="ReactVirtualized__Table__row"
        role="row"
        style="height:50px;left:0;position:absolute;top:0;width:500px;overflow:hidden;padding-right:0"
        tabindex="0"
      >
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Agent Conf"
        >
          Agent Conf
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 100px;-ms-flex:0 1 100px;-webkit-flex:0 1 100px;overflow:hidden"
          title="January 20-21, 2017"
        >
          January 20-21, 2017
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Dornbirn, Austria"
        >
          Dornbirn, Austria
        </div>
      </div>
      <div
        aria-label="row"
        class="ReactVirtualized__Table__row"
        role="row"
        style="height:50px;left:0;position:absolute;top:50px;width:500px;overflow:hidden;padding-right:0"
        tabindex="0"
      >
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="O'Reilly Velocity Conference"
        >
          O'Reilly Velocity Conference
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 100px;-ms-flex:0 1 100px;-webkit-flex:0 1 100px;overflow:hidden"
          title="January 19-22, 2017"
        >
          January 19-22, 2017
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="San Jose, CA"
        >
          San Jose, CA
        </div>
      </div>
      <div
        aria-label="row"
        class="ReactVirtualized__Table__row"
        role="row"
        style="height:50px;left:0;position:absolute;top:100px;width:500px;overflow:hidden;padding-right:0"
        tabindex="0"
      >
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Script17"
        >
          Script17
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 100px;-ms-flex:0 1 100px;-webkit-flex:0 1 100px;overflow:hidden"
          title="January 27, 2017"
        >
          January 27, 2017
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Linz, Austria"
        >
          Linz, Austria
        </div>
      </div>
      <div
        aria-label="row"
        class="ReactVirtualized__Table__row"
        role="row"
        style="height:50px;left:0;position:absolute;top:150px;width:500px;overflow:hidden;padding-right:0"
        tabindex="0"
      >
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Agile Content Conf"
        >
          Agile Content Conf
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 100px;-ms-flex:0 1 100px;-webkit-flex:0 1 100px;overflow:hidden"
          title="January 30-31, 2017"
        >
          January 30-31, 2017
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="London, UK"
        >
          London, UK
        </div>
      </div>
      <div
        aria-label="row"
        class="ReactVirtualized__Table__row"
        role="row"
        style="height:50px;left:0;position:absolute;top:200px;width:500px;overflow:hidden;padding-right:0"
        tabindex="0"
      >
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Jfokus"
        >
          Jfokus
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 100px;-ms-flex:0 1 100px;-webkit-flex:0 1 100px;overflow:hidden"
          title="February 6-8, 2017"
        >
          February 6-8, 2017
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Stockholm, Sweden"
        >
          Stockholm, Sweden
        </div>
      </div>
      <div
        aria-label="row"
        class="ReactVirtualized__Table__row"
        role="row"
        style="height:50px;left:0;position:absolute;top:250px;width:500px;overflow:hidden;padding-right:0"
        tabindex="0"
      >
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Webstock"
        >
          Webstock
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 100px;-ms-flex:0 1 100px;-webkit-flex:0 1 100px;overflow:hidden"
          title="February 13-17, 2017"
        >
          February 13-17, 2017
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Wellington, New Zealand"
        >
          Wellington, New Zealand
        </div>
      </div>
      <div
        aria-label="row"
        class="ReactVirtualized__Table__row"
        role="row"
        style="height:50px;left:0;position:absolute;top:300px;width:500px;overflow:hidden;padding-right:0"
        tabindex="0"
      >
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Sustainable UX"
        >
          Sustainable UX
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 100px;-ms-flex:0 1 100px;-webkit-flex:0 1 100px;overflow:hidden"
          title="February 16, 2017"
        >
          February 16, 2017
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Online"
        >
          Online
        </div>
      </div>
      <div
        aria-label="row"
        class="ReactVirtualized__Table__row"
        role="row"
        style="height:50px;left:0;position:absolute;top:350px;width:500px;overflow:hidden;padding-right:0"
        tabindex="0"
      >
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="The Rolling Scopes Conference"
        >
          The Rolling Scopes Conference
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 100px;-ms-flex:0 1 100px;-webkit-flex:0 1 100px;overflow:hidden"
          title="February 18-19, 2017"
        >
          February 18-19, 2017
        </div>
        <div
          class="ReactVirtualized__Table__rowColumn"
          role="gridcell"
          style="flex:0 1 200px;-ms-flex:0 1 200px;-webkit-flex:0 1 200px;overflow:hidden"
          title="Minsk, Belarus"
        >
          Minsk, Belarus
        </div>
      </div>
    </div>
  </div>
</div>
`;
